src/audio/drc/drc_math_hifi3.c

```diff
@@ -0,0 +1,263 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 Google LLC. All rights reserved.
+//
+// Author: Pin-chih Lin <johnylin@google.com>
+
+#include <sof/audio/drc/drc_math.h>
+#include <sof/math/numbers.h>
+
+#if DRC_HIFI3
+
+#include <xtensa/tie/xt_hifi3.h>
+
+/*
+ * A substitutive function of Q_MULTSR_32X32(a, b, qa, qb, qy) in HIFI manners.
+ *
+ * In AE_MULF32R, it takes the multiplication of 1.31 and 1.31 to 17.47, which means it does
+ * right-shift 15 bits. Let's consider a is .qa, b is .qb, then tmp will be .(qa+qb-15)
+ * In AE_ROUND32F48SSYM, it rounds 17.47 to 1.31, equally right-shifts 16 bits. Let's consider the
+ * output is .qy, then tmp here needs to be .(qy+16)
+ * If we set the left-shift bit from the former tmp to the latter tmp as "x", then:
+ *     (qa+qb-15) + x = (qy+16)
+ *     x = qy-qa-qb+31
+ */
+static inline ae_f32 mult(ae_f32 a, ae_f32 b, int32_t qa, int32_t qb, int32_t qy)
+{
+	ae_f64 tmp;
+
+	tmp = AE_MULF32R_LL(a, b);
+	tmp = AE_SLAA64S(tmp, qy - qa - qb + 31);
+	return AE_ROUND32F48SSYM(tmp);
+}
+
+/*
+ * Input depends on precision_x
+ * Output range [0.5, 1); regulated to Q2.30
+ */
+static inline ae_f32 rexp_fixed(ae_f32 x, int32_t precision_x, int32_t *e)
+{
+	int32_t bit = 31 - norm_int32((int32_t)x);
+
+	*e = bit - precision_x;
+
+	return AE_SRAA32(x, bit - 30);
+}
+
+/*
+ * Input is Q6.26: max 32.0
+ * Output range ~ (-inf, 1.505); regulated to Q6.26: (-32.0, 32.0)
+ */
+static inline ae_f32 log10_fixed(ae_f32 x)
+{
+	/* Coefficients obtained from:
+	 * fpminimax(log10(x), 5, [|SG...|], [1/2;sqrt(2)/2], absolute);
+	 * max err ~= 6.088e-8
+	 */
+	/* const int32_t ONE_OVER_SQRT2 = Q_CONVERT_FLOAT(0.70710678118654752f, 30);
+	 * const int32_t A5 = Q_CONVERT_FLOAT(1.131880283355712890625f, 26);
+	 * const int32_t A4 = Q_CONVERT_FLOAT(-4.258677959442138671875f, 26);
+	 * const int32_t A3 = Q_CONVERT_FLOAT(6.81631565093994140625f, 26);
+	 * const int32_t A2 = Q_CONVERT_FLOAT(-6.1185703277587890625f, 26);
+	 * const int32_t A1 = Q_CONVERT_FLOAT(3.6505267620086669921875f, 26);
+	 * const int32_t A0 = Q_CONVERT_FLOAT(-1.217894077301025390625f, 26);
+	 * const int32_t LOG10_2 = Q_CONVERT_FLOAT(0.301029995663981195214f, 26);
+	 */
+	const ae_f32 ONE_OVER_SQRT2 = 759250112; /* 1/sqrt(2) */
+	const ae_f32 A5 = 75959200;
+	const ae_f32 A4 = -285795039;
+	const ae_f32 A3 = 457435200;
+	const ae_f32 A2 = -410610303;
+	const ae_f32 A1 = 244982704;
+	const ae_f32 A0 = -81731487;
+	const ae_f32 LOG10_2 = 20201782;
+	int32_t e;
+	ae_f32 exp; /* Q7.25 */
+	ae_f32 acc; /* Q6.26 */
+	ae_f32 tmp; /* Q6.26 */
+
+	x = rexp_fixed(x, 26, &e); /* Q2.30 */
+	exp = e << 25; /* Q_CONVERT_FLOAT(e, 25) */
+
+	if ((int32_t)x > (int32_t)ONE_OVER_SQRT2) {
+		x = mult(x, ONE_OVER_SQRT2, 30, 30, 30);
+		exp = AE_ADD32(exp, 0x1000000); /* Q_CONVERT_FLOAT(0.5, 25) */
+	}
+
+	acc = mult(A5, x, 26, 30, 26);
+	acc = AE_ADD32(acc, A4);
```

Comment:
I think acc could be ae_f32x2 and you could do these with dual multiply or even MAC instructions. But you can also proceed incrementally and do the heavy lifting later.